### ISSUE TITLE
TSC's check of *.js fails property addition for a hoisted function declaration

### ISSUE THREAD

**Issue by wips** [Microsoft/TypeScript#18570]
**TypeScript Version:**  2.5.2

According to [the video](https://youtu.be/d1f6VBmWg6o?t=53m16) I'm trying to use `tsc` to validate js code. But it seems that `tsc` doesn't support function declaration hoisting on some level. When a property is added to a function which is declared below (1) the addition, it fails to pass `tsc` checking with the `error TS2339: Property 'NG_NAME' does not exist on type '() => { [x: string]: any; restrict: string; controller: (smMyService: any, config...'`. When the code is changed to (2), where function declaration precedes property addition there's no error. Both ways are correct from js point of view, so I expect 1st way doesn't fail the check.

**Code**
A directive where error happens (1):
```js
MyDirective.NG_NAME = 'smMyDirective';
export default function MyDirective() {
    return {
        restrict: 'A',
        controllerAs: 'smMyCtrl',
        // ...
    };
};
```
The way when there's no error (2):
```js
function MyDirective() {
    return {
        restrict: 'A',
        controllerAs: 'smMyCtrl',
        // ...
    };
};
MyDirective.NG_NAME = 'smMyDirective';
export default MyDirective;
```
tsconfig.json:
```json
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "outDir": "tsc-output",
    "module": "es6",
    "lib": [
      "es6",
      "dom",
      "dom.iterable"
    ],
    "suppressExcessPropertyErrors": true,
    "types": [
      "angular"
    ]
  },
  "include": [
    "src/js/boot/**/*.js"
  ],
  "exclude": [
    "node_modules",
    "**/*.test.js"
  ],
  "typeAcquisition": {
    "enable": true
  }
}
```
**Expected behavior:**
`tsc` check is passed
**Actual behavior:**
`tsc` check fails with the TS2339 error

**Comment by mhegazy** [comment]
We do a rather naive lookup to add properties to functions in a .js file, so we look function names that appeared before the property assignment. the correct solution is to defer this to post bind or do two-step binding to ensure we get the correct hoisting semantics.

**Comment by wips** [comment]
Guys, it is a bit off topic, but maybe you can suggest a workaround for this?

**Comment by mhegazy** [comment]
> Guys, it is a bit off topic, but maybe you can suggest a workaround for this?

move the function declaration higher than the property assignment.

**Comment by wips** [comment]
@mhegazy, thanks, this is exactly what I described in my initial comment, but changing the code is not what I meant by a workaround - we have 1k+ modules formatted this way and there's no intention to reformat them. Looks like we'll have to wait with tsc js checking until the bug is fixed. I would try to make a pull request with a fix, but given I've never contributed to TypeScript this may take ages.

**Comment by sandersn** [comment]
This should be fixed as of Typescript 2.9 (perhaps 2.8).
